product_barcode_constraint_per_company/models/product_product.py

```diff
@@ -17,9 +17,10 @@ class ProductProduct(models.Model):
     def _auto_init(self):
         for i, constraint in enumerate(self._sql_constraints):
             if constraint[0] == "barcode_uniq":
+                self = self.with_context(prefetch_fields=False)
```

Comment:
You are losing context on each loop iteration.

I would have done it before loop.

Moreover, you cannot shadow 'self' name. Use a new variable name like 'self_without_prefetch' or ...

@slominchar Is it wanted to have that context in super() or just for setting _sql_constraints ?